### ISSUE TITLE
Is automatic switching of datasource preferrable?

### ISSUE THREAD

**Issue by kjhazelwood** [fermi-controls/data-pool-manager#4]
I'm not sure the DPM switching data source from GETS32 to FTP when request is >=15Hz is always the desired behavior. Often, nodes seem to get bogged down if requesting FTP data for multiple devices which we often do (BLMs, BPMs). I think most of these nodes are capable of 60Hz GETS32 (I'm not certain). While the timestamps are preferable with FTP, I dont think you want to automatically switch to FTP unless there is a way for the user to disable this. Is there a DRF property that allows a >=15Hz request in which the user can tell the DPM to not use FTP? I know there is a DRF property to enable FTP, is there the opposite?

I think its also important that the ItemData.meta attribute specify the datasource so the user knows this.

What happens when the user requests a data rate, say 130Hz that is not a supported FTP class on a particular node? Does the DPM alter the request to match the nodes capabilities? If so, does the user get any notification of this? I think I would prefer the DPM error on not supported FTP class and make the user specify a supported rate.

I guess on the user end I don't see a benefit of the auto switching, I like having to be explicit so I know what I'm getting. But I can see how switching to FTP may help DPM, larger packets, less transactions?

**Comment by kjhazelwood** [comment]
I agree, most users wont care that DPM automagically switched the data source (they may care if the rate is different than requested). I do think you need to attach some information to the DPM reply so the user can query what they got. I think the easy fix for almost all my concerns is excatly what you suggest, some sort of DRF strict flag so that the DPM does its best to fulfill the request as specified and error if it cant do it.

**Comment by rneswold** [comment]
To summarize for @charlieking65:

* How hard does DPM try to match the request rate? Does it return an error if it can't use the rate or does it downgrade to a slower rate?
* Should we add a field to DRF's periodic event specification that indicates the request *must* use the rate (or return an error if it can't).

**Comment by rneswold** [comment]
After further discussion, we've decided that DPM's only responsibility in this is to make sure the request is less than or equal to the device's FTP class code rate. If it's faster, DPM will clip it to the max rate.

I will change `acsys-python` to add a status field in `ItemData`. It'll mostly be `ACNET_SUCCESS` but, in the case of downgrading a sampling rate, it'll be an ACNET warning code (to be determined.)

DPM will not verify the date is being returned at the rate requested. If the programmer determines the data is bad, they'll have to duke it out with the front-end maintainer.

**Comment by rneswold** [comment]
fermi-controls/acsys-python@6b5ed1c51bb9a286e819b26d2c0d9138baec8a58 adds the `.status` field to `ItemData`.
